Add support for JSON output

txels opened the pull request:
Implements #2 

**Describe your changes**
Adds JSON support via the new CLI flag `--json`.

**Testing performed**
Added unit tests, plus tested locally against a few particularly large virtual environments. I've refactored the `print_tree` version to take some logic out of the presentation layer and in the model layer, so that it could be reused for the new JSON output format (and potential future options).


**Sample output**

```json
[
  {
    "name": "charset-normalizer",
    "version": "2.0.7",
    "up_to_date": false,
    "current": {
      "version": "2.0.7",
      "date": "2021-10-11T21:28:25.641146+00:00",
      "since": "a month ago"
    },
    "latest": {
      "version": "2.0.8",
      "date": "2021-11-24T19:46:01.258828+00:00",
      "since": "17 hours ago"
    }
  },
  {
    "name": "setuptools",
    "version": "57.4.0",
    "up_to_date": false,
    "current": {
      "version": "57.4.0",
      "date": "2021-07-19T01:20:16.478220+00:00",
      "since": "4 months ago"
    },
    "latest": {
      "version": "59.2.0",
      "date": "2021-11-19T01:14:16.274179+00:00",
      "since": "6 days ago"
    }
  },
  {
    "name": "black",
    "version": "21.11b1",
    "up_to_date": true,
    "current": {
      "version": "21.11b1",
      "date": "2021-11-18T03:58:02.282075+00:00",
      "since": "7 days ago"
    },
    "latest": {
      "version": "21.11b1",
      "date": "2021-11-18T03:58:02.282075+00:00",
      "since": "7 days ago"
    }
  },
...
]
```

txels commented on src/pypi_changes/_print.py:
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
-                        text.append(naturaltime(now - first_release_at), "green")
-                        break
-                text.append(f" remote {last_release['version']}", "red")
-            if last_release["upload_time_iso_8601"] is not None:  # pragma: no branch
```
I'd actually like to take this refactor further and move all non-presentational logic elsewhere - a function that returns the processed dataset so that it can be fed to output formatters (rich text, json) but also consumed by 3rd party tools that use `pypi-changes` as a library.

The main point of my contribution is to improve composability of `pypi-changes` so that it can be more easily used in broader scoped tools.

gaborbernat replied:
I'm open to this.

txels replied:
I will not do this as part of this PR to expedite its acceptance.

gaborbernat commented on src/pypi_changes/__init__.py:
```diff
@@ -15,8 +15,11 @@
     distributions = collect_distributions(options)
-    info = pypi_info(distributions, options)
-    print_tree(info, options)
+    info = list(pypi_info(distributions, options))
```
I personally would prefer to keep this an iterator, and only materialize it later where it is needed to be a list.

txels replied:
Done

gaborbernat commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ __pycache__
 /src/pypi_changes/_version.py
 /src/*.egg-info
 /.*_cache
+/.venv
```
IMHO this is out of scope for this gitignore. Such configuration should be in your global git config and not in each individual project.

txels replied:
Ah I didn't know about global git ignores, that's a great tip.

txels replied:
Done

gaborbernat commented on src/pypi_changes/_cli.py:
```diff
@@ -43,6 +43,8 @@ def _define_cli_arguments() -> ArgumentParser:
     cache_help = "seconds how long requests should be cached (pass 0 to bypass the cache, -1 to cache forever)"
     parser.add_argument("--cache-duration", "-d", default=3600, type=int, help=cache_help, metavar="SEC")
 
+    parser.add_argument("--json", action="store_true", default=False, help="Show output in JSON format")
```
Would prefer to make the presentation a mutually exclusive flag combination, where the default is the current state so it's easily extensible.

txels replied:
Makes sense. How do you want to call that flag? `--output`, `--format`? And how do you want to call the current format that uses rich for console?

txels replied:
Done

gaborbernat commented on src/pypi_changes/_pkg.py:
```diff
@@ -48,6 +48,15 @@ def version(self) -> str:
+            release_info = self.info["releases"].get(self.version)
+            if release_info is not None:
+                return release_info[0] or {}
+        # return empty version info if not matching
```
Can you move this at the right side of the return below?

gaborbernat commented on src/pypi_changes/_print.py:
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
-                        break
-                text.append(f" remote {last_release['version']}", "red")
-            if last_release["upload_time_iso_8601"] is not None:  # pragma: no branch
+
+        current_release = pkg.current_release
+        current_release_at = current_release.get("upload_time_iso_8601")
+        last_release = pkg.last_release or {}
+        last_release_at = pkg.last_release_at
+
+        if current_release_at is not None:
+            text.append(" ")
+            text.append(naturaltime(now - current_release_at), "green")
+        if pkg.version != last_release.get("version"):
+            text.append(f" remote {last_release.get('version')}", "red")
```
How come we needed to change the rich layer to add the JSON presentation?

txels replied:
The output hasn't changed to the best of my knowledge. This is a refactor to try and get all of the common "data analysis and aggregation" logic outside of the presentational layer so that it can be used by the various output formats (rich, JSON, and potential future additions).

gaborbernat commented on src/pypi_changes/_print.py:
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
     rich_print(tree)
 
 
+def release_info(release: dict[str, Any] | None, now: datetime) -> dict[str, Any]:
```
I'd prefer to make the _print a package that has a two modules rich and json. And each file can only contain it's own logic. Can you do that please?

txels replied:
Sure.

txels replied:
I've done the split into `_print.json` and `_print.rich`. Not sure about what naming conventions you want to follow for these modules.

txels commented:
Thanks for the review and the gitignore tip! I'm busy and on a business trip so I don't think I'll tackle this before next week.

gaborbernat commented:
@txels do you still intend to finish this?

txels commented:
> @txels do you still intend to finish this?

Hey, yeah - I was waiting for some feedback from my latest replies (e.g. https://github.com/gaborbernat/pypi_changes/pull/5#discussion_r764793216) before going forward :grimacing: as I'd prefer to do all the work in a single sitting. I'd be happy to push this through but it's up to you. I had mixed feelings re: how favorable you were to this PR, TBH.

gaborbernat commented:
I'm happy to accept this change we just need to bikeshed on how it would operate. Do you still intend to action on the outstanding points and fix the merge conflicts?

txels commented:
I've rebased and done all the changes you suggested @gaborbernat - I've also fixed an incorrect mock that made `pytest` consistently fail locally.

I'm bumping into this issue in CI, where coverage report is failing with a generated file: 

```
packages/coverage/report.py:87: CoverageWarning: Couldn't parse '/Users/runner/work/pypi_changes/pypi_changes/src/pypi_changes/_version.py': No source for code: '/Users/runner/work/pypi_changes/pypi_changes/src/pypi_changes/_version.py'. (couldnt-parse)
coverage._warn(msg, slug="couldnt-parse")
```

...not sure what is the best way to proceed, as I don't understand why this is an issue on my branch in particular.

txels commented:
Still one open pep8 issue `src/pypi_changes/_print/json.py:7:22: SC200 Possibly misspelt word: 'naturaldelta'` - how do you want to go about it?

txels commented:
OK done for today, let me know what you think of the state of this so far and next steps.

txels commented:
@gaborbernat this is now green and conflict free. I don't think I want to spend much more time on this, it's already long gone into a zone of "diminishing returns" for me.

gaborbernat commented on src/pypi_changes/_print/json.py:
```diff
@@ -0,0 +1,47 @@
+from datetime import datetime, timezone
+from typing import Any, Iterable
+
+from humanize import naturaldelta  # noqa: SC200
```
Instead, you should add those to the whitelist.txt at the root of the project?

gaborbernat commented on src/pypi_changes/_print/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from .._pkg import Package
+
+
+class Reversor:
```
This probably could/should be private so named _Reversor? or nested within the method below?

txels replied:
This was already in the original code.

txels replied:
Done

gaborbernat commented on tests/test_print rich.py:
```diff
@@ -73,8 +73,8 @@ def test_print_alphabetical(capsys: CaptureFixture[str], option_simple: Options,
     assert output[0].startswith("🐍 Distributions within")
     assert output[1] == sys.executable or output[0].endswith(sys.executable)
     assert output[-4:] == [
-        "├── a 1 remote 2 a month ago",
```
Wonder why this change?

txels replied:
The original issue was that you were using `naturaltime` whose type signature accepts `datetime|int`, which mypy was flagging. Looking into the `humanize` library, I realised that we should move to use `naturaldelta` that has the right signature (takes a `timedelta`). However, that function doesn't include `ago` in the output.

txels replied:
TBH looking at the code of `humanize` the typing is inconsistent, you may consider to contribute a patch there.

https://github.com/python-humanize/humanize/blob/main/src/humanize/time.py#L208 <<< `naturaltime` signature doesn't take a timedelta - however, internally it accepts it https://github.com/python-humanize/humanize/blob/main/src/humanize/time.py#L238

gaborbernat reviewed (changes requested):
Otherwise looks good 👍

gaborbernat reviewed (approved):
![](https://media.tenor.co/images/b05a23c5855c35db77f39915f3ec3cbe/tenor.gif)

gaborbernat commented:
Thanks for sticking through it! Now should be released.
